### PR TITLE
[MHA] Implement MIOPEN_BACKEND_OPERATION_MATMUL_DESCRIPTOR

### LINKED ISSUE #2840: [MHA] Implement batch matrix multiplication descriptor in Graph API
The following descriptors are to be implemented.

MIOPEN_BACKEND_MATMUL_DESCRIPTOR
- [x] Builder and tests
- [x] C functions and tests

MIOPEN_BACKEND_OPERATION_MATMUL_DESCRIPTOR
- [x] Builder and tests
- [x] C functions and tests

### PR DESCRIPTION
**Pull request by kikimych**
Implement one #2840 issue

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -125,7 +125,8 @@ endif()
 ############################################################
 # OPTION - MIOpen Backend
 # - OpenCL
-# - HIP
+# - HI
+message("HAS_HIP == " ${HAS_HIP})
```
Comment (atamazov) [inline CMakeLists.txt]: Maybe this change is not necessary? Or use VERBOSE or DEBUG message mode.

File: CMakeLists.txt [hunk 2]
```diff
@@ -282,6 +283,7 @@ if(HAS_HIP)
 else()
     # CXX compiler is not HIP compiler, let's analyze HIP version.
     set(MIOPEN_HIP_COMPILER_HAS_OPTION_OFFLOAD_UNIFORM_BLOCK Off)
+    message(MIOPEN_hip_VERSION_FLAT="${MIOPEN_hip_VERSION_FLAT}")
```
Comment (atamazov) [inline CMakeLists.txt]: We print HIP version at line 220.

File: src/graphapi/matmul.cpp [hunk 3]
```diff
@@ -104,5 +104,254 @@ void BackendMatmulDescriptor::getAttribute(miopenBackendAttributeName_t attribut
...
 }
 
+OperationMatmul OperationMatmulBuilder::build()
+{
+    if(!aSet || !bSet || !cSet || !matmulSet)
+        MIOPEN_THROW(miopenStatusBadParm);
+
+    if(!(mOperationMatmul.mA->getDimensions().size() == mOperationMatmul.mB->getDimensions().size() == mOperationMatmul.mC->getDimensions().size()))
+        MIOPEN_THROW(miopenStatusBadParm);
+    int Am = mOperationMatmul.mA->getDimensions().end()[-2];
```
Comment (atamazov) [inline src/graphapi/matmul.cpp]: [Q] Do we guarantee that `mOperationMatmul.mA->getDimensions().end()[-2]` exists?

**Review by atamazov — CHANGES_REQUESTED** [review]
Perhaps it's too early to review and this is a draft? If so, the please convert this PR to draft. Thanks.

File: CMakeLists.txt (hunk 1 above)
Comment (kikimych) [reply CMakeLists.txt]: accidentally touched CMakeLists.txt

File: CMakeLists.txt (hunk 2 above)
Comment (kikimych) [reply CMakeLists.txt]: same as above

File: src/graphapi/matmul.cpp (hunk 3 above)
Comment (kikimych) [reply src/graphapi/matmul.cpp]: Fixed in setters, added tests

**Comment by atamazov** [comment]
[Attribution] @junliume @JehandadKhan
- https://github.com/ROCm/MIOpen/labels/enhancement
- https://github.com/ROCm/MIOpen/labels/urgency_high
- https://github.com/ROCm/MIOpen/labels/value_high
- Proposed reviewers:
  - @scerzh 
  - @Vsevolod1983 
  - @amberhassaan 
  - @CAHEK7 
  - @DrizztDoUrden

File: CMakeLists.txt (hunk 1 above)
Comment (atamazov) [reply CMakeLists.txt]: @kikimych Please do not forget to revert ;)

File: src/graphapi/matmul.cpp (hunk 3 above)
Comment (atamazov) [reply src/graphapi/matmul.cpp]: [resolved]

File: src/graphapi/matmul.cpp [hunk 4]
```diff
@@ -26,6 +26,7 @@
 
 #include <miopen/errors.hpp>
 #include <miopen/graphapi/matmul.hpp>
+#include <iostream>
```
Comment (scerzh) [inline src/graphapi/matmul.cpp]:
```suggestion
```
Seems to be added by an accident or for debugging only.

File: src/include/miopen/graphapi/rng.hpp [hunk 5]
```diff
@@ -115,7 +115,6 @@ class BackendRngDescriptor : public BackendDescriptor
 private:
     RngBuilder mBuilder;
     Rng mRng;
-
 public:
```
Comment (scerzh) [inline src/include/miopen/graphapi/rng.hpp]:
```suggestion

public:
```
Triggers clang-format violation. Seems to be changed by an accident. Could you undo this chunk, please?

File: src/include/miopen/graphapi/matmul.hpp [hunk 6]
```diff
@@ -82,5 +85,113 @@ class BackendMatmulDescriptor : public BackendDescriptor
...
+};
+
+class OperationMatmulBuilder
+{
+private:
+    OperationMatmul mOperationMatmul;
+    bool aSet      = false;
+    bool bSet      = false;
+    bool cSet      = false;
+    bool matmulSet = false;
```
Comment (scerzh) [inline src/include/miopen/graphapi/matmul.hpp]:
```suggestion
    bool mASet      = false;
    bool mBSet      = false;
    bool mCSet      = false;
    bool mMatmulSet = false;
```
Colleagues insist on sticking to our new naming rules.

File: src/graphapi/matmul.cpp (hunk 3 above)
Comment (scerzh) [reply src/graphapi/matmul.cpp]:
```suggestion
    int Am = mOperationMatmul.mA->getDimensions().begin()[aDimensionsCount-2];
```
`cppcheck` doesn't allow to dereference `end()` iterators.

File: src/graphapi/matmul.cpp [hunk 7]
```diff
@@ -104,5 +105,310 @@ void BackendMatmulDescriptor::getAttribute(miopenBackendAttributeName_t attribut
...
+        MIOPEN_THROW(miopenStatusBadParm);
+
+    int Am = mOperationMatmul.mA->getDimensions().end()[-2];
+    int An = mOperationMatmul.mA->getDimensions().end()[-1];
+
+    int Bn = mOperationMatmul.mB->getDimensions().end()[-2];
+    int Bk = mOperationMatmul.mB->getDimensions().end()[-1];
+
+    int Cm = mOperationMatmul.mB->getDimensions().end()[-2];
+    int Ck = mOperationMatmul.mB->getDimensions().end()[-1];
```
Comment (scerzh) [inline src/graphapi/matmul.cpp]:
```suggestion
    int An = mOperationMatmul.mA->getDimensions().begin()[aDimensionsCount-1];

    int Bn = mOperationMatmul.mB->getDimensions().begin()[bDimensionsCount-2];
    int Bk = mOperationMatmul.mB->getDimensions().begin()[bDimensionsCount-1];

    int Cm = mOperationMatmul.mC->getDimensions().begin()[cDimensionsCount-2];
    int Ck = mOperationMatmul.mC->getDimensions().begin()[cDimensionsCount-1];
```
Same as above: `cppcheck` doesn't allow to dereference `end()` iterators.
And more serious, could you check the last 2 lines: should there be `mB` or `mC`?

File: src/graphapi/matmul.cpp [hunk 8]
```diff
@@ -104,5 +105,310 @@ void BackendMatmulDescriptor::getAttribute(miopenBackendAttributeName_t attribut
...
+}
+
+void BackendOperationMatmulDescriptor::finalize()
+{
+    if(mFinalized || !deref(mA).isFinalized() || !deref(mB).isFinalized() ||
+       !deref(mC).isFinalized() || !deref(mGemmMOverride).isFinalized() ||
+       !deref(mGemmNOverride).isFinalized() || !deref(mGemmKOverride).isFinalized())
+    {
+        MIOPEN_THROW(miopenStatusNotInitialized);
+    }
```
Comment (scerzh) [inline src/graphapi/matmul.cpp]: For invalid attributes we throw `miopenStatusBadParm`.

By the time of execution of this function the fulfillment of `isFinalized()` requirement was already checked in `setAttribute()`, I believe, there is no need of redoing it here.

At first I wanted to omit `isFinalized()` checks:

```suggestion
    if(mFinalized)
    {
        MIOPEN_THROW(miopenStatusNotInitialized);
    }
```

but upon reviewing cuDNN's docs [cudnnbBackendFinalize()](https://docs.nvidia.com/deeplearning/cudnn/latest/api/cudnn-graph-library.html#cudnnbackendfinalize) once again I think we need to simply return here:

```suggestion
    if(mFinalized)
    {
        return;
    }
```

So I was wrong in `throw`ing in all such cases in my PRs. Could you also take a look at the cuDNN documentation and give your opinion, please?

**Review by scerzh — CHANGES_REQUESTED** [review]
Good. There are a few `cppcheck` and `clang-format` issues and a question about the behavior of `finalize()`

File: src/graphapi/matmul.cpp [hunk 9]
```diff
@@ -104,5 +105,310 @@ void BackendMatmulDescriptor::getAttribute(miopenBackendAttributeName_t attribut
...
+    int Bk = mOperationMatmul.mB->getDimensions().end()[-1];
+
+    int Cm = mOperationMatmul.mB->getDimensions().end()[-2];
+    int Ck = mOperationMatmul.mB->getDimensions().end()[-1];
+
+    if(Am != Cm || An != Bn || Bk != Ck)
+        MIOPEN_THROW(miopenStatusBadParm);
+
+    // TODO: need broadcsat checks
+    //
```
Comment (scerzh) [inline src/graphapi/matmul.cpp]: I did the broadcast checks for pointwise descriptor here [pointwise.cpp](https://github.com/ROCm/MIOpen/blob/0334279c5214329acd8f7dfe20b31182e17c6bdb/src/graphapi/pointwise.cpp#L686-L712). Could you check whether it fits your needs? If so, reuse the template and move it to one of the common headers, probably `graphapi.hpp`.

\cc @CAHEK7 @DrizztDoUrden

File: src/graphapi/matmul.cpp (hunk 8 above)
Comment (kikimych) [reply src/graphapi/matmul.cpp]: It's check for double finalization

Comment (kikimych) [reply src/graphapi/matmul.cpp]: It should return CUDNN_STATUS_BAD_PARAMS

Comment (kikimych) [reply src/graphapi/matmul.cpp]: > It should return CUDNN_STATUS_BAD_PARAMS

User can check result of the function call via std::optional<>, status code, catching exception or input-output parameters. At least one way should exist. So, throwing exception is better then return void, but returning status code does not call std::terminate() in worst case scenario. Crashing application from library is bad idea

**Comment by junliume** [comment]
Some issues:
```
[2024-04-19T17:49:54.041Z] [100%] Linking CXX executable ../../bin/miopen_gtest

[2024-04-19T17:49:54.041Z] ld.lld: error: duplicate symbol: miopen::graphapi::operator<<(std::ostream&, std::variant<long, miopen::graphapi::Tensor*>)

[2024-04-19T17:49:54.041Z] >>> defined at graphapi_operation_matmul.cpp

[2024-04-19T17:49:54.041Z] >>>            CMakeFiles/miopen_gtest.dir/graphapi_operation_matmul.cpp.o:(miopen::graphapi::operator<<(std::ostream&, std::variant<long, miopen::graphapi::Tensor*>))

[2024-04-19T17:49:54.041Z] >>> defined at graphapi_operation_rng.cpp

[2024-04-19T17:49:54.041Z] >>>            CMakeFiles/miopen_gtest.dir/graphapi_operation_rng.cpp.o:(.text+0x0)
```

File: src/graphapi/matmul.cpp (hunk 7 above)
Comment (kikimych) [reply src/graphapi/matmul.cpp]: mC, thanks

File: src/graphapi/matmul.cpp (hunk 9 above)
Comment (kikimych) [reply src/graphapi/matmul.cpp]: fixed

File: src/graphapi/matmul.cpp (hunk 4 above)
Comment (kikimych) [reply src/graphapi/matmul.cpp]: fixed

File: src/include/miopen/graphapi/rng.hpp (hunk 5 above)
Comment (kikimych) [reply src/include/miopen/graphapi/rng.hpp]: fixed

File: src/include/miopen/graphapi/matmul.hpp (hunk 6 above)
Comment (kikimych) [reply src/include/miopen/graphapi/matmul.hpp]: fixed

File: src/graphapi/matmul.cpp (hunk 7 above)
Comment (kikimych) [reply src/graphapi/matmul.cpp]: fixed

File: src/graphapi/matmul.cpp (hunk 8 above)
Comment (atamazov) [reply src/graphapi/matmul.cpp]: @kikimych 
> So, throwing exception is better then return void, but returning status code does not call std::terminate() in worst case scenario.

~The initial idea was to format all the user's hard disks, but finally~ we ~decided to~ catch exceptions at API wrapper level and return error code. Please check yourself.

Comment (kikimych) [reply src/graphapi/matmul.cpp]: got it

**Review by scerzh — APPROVED** [review]
LGTM

**Comment by junliume** [comment]
CI restarted for this PR.

**Review by junliume — APPROVED** [review]

**Pull request merged**